architecture-decision-records/008-greenhouse-controller-lifecycle.md

```diff
@@ -0,0 +1,132 @@
+# 008 Greenhouse Controller Lifecycle
+
+- Status: proposed
+- Deciders: Ivo Gosemann, Uwe Mayer, David Gogl, Abhijith Ravindra
+- Date: 2024-10-18
+- Tags: greenhouse
+- Technical Story: [greenhouse#414](https://github.com/cloudoperators/greenhouse/issues/414)
+
+## Context and Problem Statement
+
+Greenhouse contains multiple controllers for custom resources such as `Organization`, `Cluster`, `Plugin`, etc. 
+
+These controllers need a unified approach for their reconciliation lifecycle.
+
+Therefore, this ADR addresses the following concerns:
+
+- A standard reconcile interface that should be adopted by existing controllers and all new controllers.
+- A common place to set the status of the resource as proposed by the ADR [004 Greenhouse Resource States](004-greenhouse-resource-status-reporting.md).
+
+
+## Decision Drivers
+
+- Uniformity:
+    * All controllers implement the same interface for reconciliation.
+    * Reducing code duplication
+
+- Expandability:
+    * Existing controllers should be easily able to adopt the interface.
+
+- Ease of use:
+    * New controllers should be able to implement the interface with minimal effort.
+
+- Simplicity:
+    * Controllers adopting the interface do not need to implement the reconciliation logic themselves.
+    * `Create / Update` and `Delete` logic is separated.
+
+## Decision Outcome
+
+As described in the issue [greenhouse#414](https://github.com/cloudoperators/greenhouse/issues/414) we will introduce a `Reconciler` interface that controllers can implement.
+
+## Reconciler and RuntimeObject Interface
+
+The RuntimeObject interface allows the `Reconciler` to work with any CR object in a generic way, allowing the reconciler to access `DeepCopyObject` and the `ObjectMeta` of the CR object - `GetNamespace()`, `GetName()`, etc.
+
+https://github.com/cloudoperators/greenhouse/blob/bb57e128014102f963c9879ef78af80bc1820bd4/pkg/lifecycle/reconcile.go#L48-L56
+
+The `Reconciler` interface will have the following methods `EnsureCreated` and `EnsureDeleted` that the calling controller should implement.
+
+https://github.com/cloudoperators/greenhouse/blob/bb57e128014102f963c9879ef78af80bc1820bd4/pkg/lifecycle/reconcile.go#L58-L62
+
+`Reconcile` - is a generic function that is used to reconcile the state of a resource
+It standardizes the reconciliation loop and provides a common way to set finalizers, remove finalizers, and update the status of the resource
```

Comment:
Could you mention that the original object is added to the `context.Context` as well as part of the `Reconcile`?